### PR TITLE
Incubator.TextField - should float with defaultValue

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
Incubator.TextField - FloatingPlaceholder should float with defaultValue
Adding `shouldFloat` and fixing its logic

WOAUILIB-3123

Handling edge cases, using the below `PlaygroundScreen`:
1. Switching between `TextField`s should:
    a. float\un-float if both are empty
    b. float\un-float on an empty one and remain un-floated on one with a `value` \ `defaultValue`
2. Changing `defaultValue` should affect the floating status
3. Deleting the `defaultValue` (to an empty `value`) should un-float

```
import React, {Component} from 'react';
import {View, Text, Card, Incubator, Button} from 'react-native-ui-lib';

export default class PlaygroundScreen extends Component {
  state = {
    defaultValue: 'initial value'
  };
  render() {
    return (
      <View bg-grey80 flex padding-20>
        <View marginT-20>
          <Incubator.TextField
            migrate
            marginH-s5
            placeholder={'placeholder'}
            defaultValue={this.state.defaultValue}
            floatingPlaceholder
            floatOnFocus
          />
          <Incubator.TextField migrate marginH-s5 placeholder={'placeholder'} floatingPlaceholder floatOnFocus/>
        </View>
        <Card height={100} center padding-20>
          <Text text50>Playground Screen</Text>
        </Card>
        <View flex center>
          <Button
            marginV-20
            label="Button"
            onPress={() => {
              this.setState({defaultValue: undefined});
            }}
          />
        </View>
      </View>
    );
  }
}
```

## Changelog
Incubator.TextField - FloatingPlaceholder should float with defaultValue

### CODE REVIEW

File: src/incubator/TextField/FloatingPlaceholder.tsx [hunk 1]
```diff
@@ -18,24 +19,29 @@ const FloatingPlaceholder = ({
...
   testID
 }: FloatingPlaceholderProps) => {
   const context = useContext(FieldContext);
   const [placeholderOffset, setPlaceholderOffset] = useState({
     top: 0,
     left: 0
   });
-  const animation = useRef(new Animated.Value(Number((floatOnFocus && context.isFocused) || context.hasValue))).current;
+
+  const shouldFloat = useMemo(() => {
```
Comment (ethanshar) [inline src/incubator/TextField/FloatingPlaceholder.tsx:31]: There's no need to memoize primitive values (booleans, strings, etc...) 😬

Comment (M-i-k-e-l) [reply src/incubator/TextField/FloatingPlaceholder.tsx:31]: Wouldn't that break the `useDidUpdate` dependencies? Unless I put the dependencies from here, which is a little weird.

Comment (M-i-k-e-l) [reply src/incubator/TextField/FloatingPlaceholder.tsx:31]: I think you're right but I found a bug :/

File: src/incubator/TextField/types.ts [hunk 2]
```diff
@@ -86,6 +86,7 @@ export interface FloatingPlaceholderProps {
   floatOnFocus?: boolean;
   validationMessagePosition?: ValidationMessagePosition;
   extraOffset?: number;
+  defaultValue?: TextInputProps['defaultValue'];
```
Comment (Inbal-Tish) [inline src/incubator/TextField/types.ts:89]: 'TextInputProps' are the OLD TextField's prop... why use them in the incubator?

File: src/incubator/TextField/index.tsx [hunk 3]
```diff
@@ -125,6 +125,7 @@ const TextField = (props: InternalTextFieldProps) => {
           <View flexG /* flex row */>
             {floatingPlaceholder && (
               <FloatingPlaceholder
+                defaultValue={others.defaultValue}
```
Comment (Inbal-Tish) [inline src/incubator/TextField/index.tsx:128]: Why not pass the input's 'value'? Why do we need a separate prop?

File: src/incubator/TextField/FloatingPlaceholder.tsx [hunk 4]
```diff
@@ -18,24 +19,27 @@ const FloatingPlaceholder = ({
   floatOnFocus,
   validationMessagePosition,
   extraOffset = 0,
+  defaultValue,
```
Comment (Inbal-Tish) [inline src/incubator/TextField/FloatingPlaceholder.tsx:22]: The FlotingPlaceholder only needs to know only if to float or not, why doesn't the TextField component controls it? Like passing in 'shouldFloat' prop as boolean that will be updated by the state of the input's value?

File: src/incubator/TextField/types.ts (hunk 2 above)
Comment (M-i-k-e-l) [reply src/incubator/TextField/types.ts:89]: No, it's RN's props

File: src/incubator/TextField/index.tsx (hunk 3 above)
Comment (M-i-k-e-l) [reply src/incubator/TextField/index.tsx:128]: This is a prop from RN which we inherit and the user that opened the bug is using

File: src/incubator/TextField/FloatingPlaceholder.tsx (hunk 4 above)
Comment (M-i-k-e-l) [reply src/incubator/TextField/FloatingPlaceholder.tsx:22]: You're saying that `shouldFloat` should be coming from somewhere else? Maybe the `Presenter` like in `ValidationMessage`?

Comment (Inbal-Tish) [reply src/incubator/TextField/FloatingPlaceholder.tsx:22]: Yes. I think the floating placeholder should only get a boolean and the logic will be in the component that uses it

File: src/incubator/TextField/types.ts (hunk 2 above)
Comment (Inbal-Tish) [reply src/incubator/TextField/types.ts:89]: Ok...

**Review by Inbal-Tish — APPROVED** [review]

**Pull request merged**